icu4j/main/tests/core/src/com/ibm/icu/dev/test/message2/MessageFormat2Test.java

```diff
@@ -0,0 +1,396 @@
+                .setFunctionRegistry(registry)
+                .setPattern(message)
+                .build();
+
+        final int maxCount = 10;
+        for (int count = 0; count < maxCount; count++) {
+            assertEquals("cached formatter",
+                    "Testing " + count + "°C.",
+                    mf2.formatToString(Args.of("count", count, "unit", "C")));
+            assertEquals("cached formatter",
+                    "Testing " + count + "°F.",
+                    mf2.formatToString(Args.of("count", count, "unit", "F")));
+        }
+
+        // Check that the constructor was only called once,
+        // and the formatter as many times as the public call to format.
+        assertEquals("cached formatter", 1, counter.constructCount);
+        assertEquals("cached formatter", maxCount * 2, counter.formatCount);
+        assertEquals("cached formatter", 1, counter.fFormatterCount);
+        assertEquals("cached formatter", 1, counter.cFormatterCount);
+
+        // Check that the skeleton is respected
+        assertEquals("cached formatter",
+                "Testing 12°C.",
+                mf2.formatToString(Args.of("count", 12, "unit", "C")));
+        assertEquals("cached formatter",
+                "Testing 12.50°F.",
+                mf2.formatToString(Args.of("count", 12.5, "unit", "F")));
+        assertEquals("cached formatter",
+                "Testing 12.54°C.",
+                mf2.formatToString(Args.of("count", 12.54, "unit", "C")));
+        assertEquals("cached formatter",
+                "Testing 12.54°F.",
+                mf2.formatToString(Args.of("count", 12.54321, "unit", "F")));
+
+        message = "{Testing {$count :temp unit=$unit skeleton=(.0/w)}.}";
+        mf2 = MessageFormatter.builder()
+                .setFunctionRegistry(registry)
+                .setPattern(message)
+                .build();
+        // Check that the skeleton is respected
+        assertEquals("cached formatter",
+                "Testing 12°C.",
+                mf2.formatToString(Args.of("count", 12, "unit", "C")));
+        assertEquals("cached formatter",
+                "Testing 12.5°F.",
+                mf2.formatToString(Args.of("count", 12.5, "unit", "F")));
+        assertEquals("cached formatter",
+                "Testing 12.5°C.",
+                mf2.formatToString(Args.of("count", 12.54, "unit", "C")));
+        assertEquals("cached formatter",
+                "Testing 12.5°F.",
+                mf2.formatToString(Args.of("count", 12.54321, "unit", "F")));
+    }
+
+    @Test
+    public void testPluralWithOffset() {
+        String message = ""
+                + "match {$count :plural offset=2}\n"
+                + " when 1 {Anna}\n"
+                + " when 2 {Anna and Bob}\n"
+                + " when one {Anna, Bob, and {$count :number offset=2} other guest}\n"
+                + " when * {Anna, Bob, and {$count :number offset=2} other guests}\n";
```

Comment:
question: Is there no syntax at all like MF `#` to get the output of the formatted `$count`?
Users should not have to repeat the placeholder details like this.

Reply:
There is no need to do it.
The `#` is replaced by the locale variables.
The next test (`testPluralWithOffsetAndLocalVar`) checks that it works:
```
let $foo = {$count :number offset=2}
match {$foo :plural}
  when 1 {Anna}
  when 2 {Anna and Bob}
  when one {Anna, Bob, and {$foo} other guest}
  when * {Anna, Bob, and {$foo} other guests}
```

It is unclear from the spec if the placeholders in the messages would need to also inherit from the condition in the match.
It was proposed, but I argued against, thinking it would have too many side-effects (it would be as if every condition in the match would declare a "fake" local variable)

We can revisit, if you think we should do that.

Reply:
👍 tnx